Add full autocompletion capabilities to language server

brandonlangley opened the pull request:
Add parameter, block, and value autocompletion to language server

NEAMS Workbench demonstrating some capabilities on this branch:

https://github.com/idaholab/moose/assets/21372118/253792ce-883c-480b-b68c-176586d23eed

refs #22766

brandonlangley committed:
Add parameter, block, and value autocompletion to language server

refs #22766

brandonlangley commented:
@loganharbour @dschwen -

This adds full autocompletion capabilities to the MOOSE language server.

It is all working really well and is super snappy in the NEAMS Workbench.

I had to make two minor update in the Workbench code to:

 - use LSP documentation field rather than detail field for descriptions

 - truncate long descriptions in the autocomplete list for cleaner display

These changes haven't made a Workbench release yet, but they will soon.

This uses syntax paths and object types (or defaults) of the cursor position.

And it basically handles autocomplete in three main high level categories:

 - input parameters

   - starting with parser global parameters (currently active / inactive)

   - plus all action parameters for the object context path of the cursor

   - plus all object parameters for object type if it is valid for syntax

   - building an input default for each parameter and using descriptions

   - indicating in description when a parameter is required in the input

   - removing requirement via object_params_set_by_action as needed

   - filtering parameters that are deprecated, private, or already exist

 - blocks / subblocks

   - top level blocks when at document root level outside of all blocks

   - all '*' subblocks and named subblocks from within any block context

   - inserting opening and closing brace syntax of blocks on completion

 - parameter values

   - adding existing subblock children names for active / inactive values

   - adding 'true' and 'false' string values if parameter type is boolean

   - adding valid enums and descriptions if parameter is an enum type

   - adding all verified object names if (non-enum) parameter is 'type'

   - adding input lookup values if parameter type has associated syntax

     - i.e., names of Functions, Variables, Materials, Postprocessors

I documented the logic well with comments so intentions should be clear.

I also added these autocompletion scenarios to the MooseServer unit test.

Could I ask for a code review to get this merged for the NEAMS milestone?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1642654/) on 289b17c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24927/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1642672/) on 289b17c wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4e81f9</th>
      <th colspan=4>#24927 289b17</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.58%</td>
      <td>85.60%</td>
      <td>+0.02%</td>
      <td>93.23%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89461</td>
      <td>89689</td>
      <td>+228</td>
      <td>234</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15076</td>
      <td>15092</td>
      <td>+16</td>
      <td>17</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24927/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24927/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24927/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1654601/) on 8c30d2d wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 34.0%)

brandonlangley commented:
@dschwen @loganharbour -

Thanks a lot for the update in https://github.com/idaholab/moose/pull/25027 that removes the `_type` parameter.

I'll rebase this branch on that change in the next day or so, and I'll also simplify this [condition](https://github.com/brandonlangley/moose/blob/moose_server_autocomplete_22766/framework/src/base/MooseServer.C#L460)

```cpp
if (deprecated || is_private || existing_params.count(param_name) || param_name == "_type")
  continue;
```

in my autocomplete parameter filtering to just be the following, which will now be equivalent.

```cpp
if (deprecated || is_private || existing_params.count(param_name))
  continue;
```

Additionally, I just made this video of Workbench using the language server from this branch for autocompletion.

It doesn't include all features added here or anything but should give an idea of how some of this logic plays out.

In the bottom validation panel, you'll also see the input diagnostics that the server reports as the file is modified.

https://github.com/idaholab/moose/assets/21372118/f5c86fad-dc27-4a8b-876b-8e25269a0245

brandonlangley commented:
@dschwen @loganharbour -

I rebased this to include the `_type` removal from https://github.com/idaholab/moose/pull/25027 and also removed that condition from my autocomplete filtering.

Let me know if you have any issues or concerns with this PR or if you're slammed and I should ask someone else to review.

dschwen commented on unit/src/MooseServerTest.C:
```diff
@@ -663,7 +1141,164 @@ TEST_F(MooseServerTest, DocumentCompletionRequest)
+
+  EXPECT_EQ(request_id, response_id);
+
+  EXPECT_EQ(4, (int)completions_array.size());
```
```suggestion
  EXPECT_EQ(4u, completions_array.size());
```

brandonlangley replied:
Thanks, I pushed a commit that address this.

dschwen reviewed (commented):
This looks mostly fine. I have some minor comments that you may wish to dismiss as they are mostly cosmetic :-)

brandonlangley committed:
Use std::optional to fix empty text string parsing by MooseServer

refs #22766

brandonlangley committed:
Update MooseServer to address style and standards review feedback

refs #22766

brandonlangley commented:
@dschwen -

Thanks a lot for the review feedback and suggestions, I just pushed a [commit](https://github.com/idaholab/moose/pull/24927/commits/145e8369fd68a721a37a5812b1a4bc43f0d3d3b7) that addresses all of your comments.

I applied those same style changes to the rest of the MooseServer code as well, which removed a lot of blank lines.

I agree that clustering sections of logic like this and using blank lines more sparingly does improve code readability.

I'm going to keep these notes in mind moving forward and continue this style when writing any future MOOSE code.

- - - - -

I added one more small [commit](https://github.com/idaholab/moose/pull/24927/commits/7bb7d0c308c4fc6eb661f7a77590716480fc2927) that fixes an issue observed if the MooseServer app tries to parse an empty string.

The Parser's optional text string parameter was defaulted as empty which was meant to indicate it wasn't provided.

This didn't work if a client passed the server an actual empty input as the Parser would fall back to parse from disk.

Strings don't have a natural sentinel value indicating absence, so I fixed this by making the parameter std::optional.

- - - - -

I went ahead and rebased these changes on top of devel and I think this should be good for final review and merge.

If you get a chance could you take a look at these two newest commits and let me know if you approve everything?

brandonlangley commented on framework/include/base/MooseServer.h:
```diff
@@ -68,6 +72,99 @@ class MooseServer : public wasp::lsp::ServerImpl
                                      int line,
                                      int character);
 
+  /** get names of parameters and subblocks specified in given input node
```
Thanks, I pushed a commit that address this.

brandonlangley commented on framework/src/base/MooseServer.C:
```diff
@@ -168,18 +188,675 @@ MooseServer::updateDocumentTextChanges(const std::string & replacement_text,
+
+  // set used to gather all parameters valid from object context of request
+
+  InputParameters valid_params = emptyInputParameters();
```
Thanks, I pushed a commit that address this.

brandonlangley commented on framework/src/base/MooseServer.C:
```diff
@@ -168,18 +188,675 @@ MooseServer::updateDocumentTextChanges(const std::string & replacement_text,
+
+      // walk backward through syntax path adding subblock names to parents
+
+      for (size_t last_slash; (last_slash = syntax_path.find_last_of("/")) != std::string::npos;)
```
Thanks, I pushed a commit that address this.

brandonlangley commented on unit/src/MooseServerTest.C:
```diff
@@ -83,6 +83,96 @@ protected:
+    };
+
+    std::vector<CompletionInfo> completions;
+    size_t max_label = 0;
```
Thanks, I pushed a commit that address this.

loganharbour commented:
I've verified that @dschwen's comments are addressed.

The logic with `optional` looks better.
